nexus/src/db/datastore.rs

```diff
@@ -2845,6 +2848,115 @@ impl DataStore {
         Ok(())
     }
 
+    // Role assignments
+
+    /// Fetches all of the role assignments for the specified resource
+    ///
+    /// This function is generic over all resources that can accept roles (e.g.,
+    /// Fleet, Silo, Organization, etc.).
+    // TODO-scalability In an ideal world, this would be paginated.  The impact
+    // is mitigated because we cap the number of role assignments per resource
+    // pretty tightly.
+    pub async fn role_assignment_fetch_all<
+        T: authz::ApiResourceWithRoles + Clone,
+    >(
+        &self,
+        opctx: &OpContext,
+        authz_resource: &T,
+    ) -> ListResultVec<db::model::RoleAssignment> {
+        opctx.authorize(authz::Action::ReadPolicy, authz_resource).await?;
+        let resource_type = authz_resource.resource_type();
+        let resource_id = authz_resource.resource_id();
+        use db::schema::role_assignment::dsl;
+        dsl::role_assignment
+            .filter(dsl::resource_type.eq(resource_type.to_string()))
+            .filter(dsl::resource_id.eq(resource_id))
+            .order(dsl::role_name.asc())
+            .then_order_by(dsl::identity_id.asc())
+            .select(RoleAssignment::as_select())
+            .load_async::<RoleAssignment>(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
+    }
+
+    /// Removes all existing role assignments on `authz_resource` and adds those
+    /// specified by `new_assignments`
+    ///
+    /// The expectation is that the caller will have just fetched the role
+    /// assignments, modified them, and is giving us the complete new list.
+    ///
+    /// This function is generic over all resources that can accept roles (e.g.,
+    /// Fleet, Silo, Organization, etc.).
+    // TODO-correctness As with the rest of the API, we're lacking an ability
+    // for an ETag precondition check here.
```

Comment:
I suspect there's a lot of work here but I don't know.  I feel like several times over the last year I've thought "ugh, <thing I'm working on> is gonna get complicated when we implement etags".  But thinking out loud:

- We need to decide how to generate the etag for every kind of resource.  Is it random (and updated with every modification) or content-based?  If content-based, is it stored in the database or calculated when needed?  If calculated when needed, it'd be nice if we could do it in SQL so that you can calculate it and check it within a SQL query (see below).  What fields go into it?  How do we specify that on a per-resource basis?  How do we keep it up to date if you're doing something like `update ... set ... where <condition that matches 10 rows>`?
- We need to plumb it through on "list" and "get" for everything.  Gotta figure out where to store it...if it's stored in
the database, great.  If not, how do we pass it around?  Maybe it goes into the `authz` object?  That might sound weird but the `authz` object basically represents a handle to a specific object, regardless of how it was looked up or what it's called now.  It makes sense that would also be locked to a specific revision of that object.
- We need to modify all the "update" datastore functions so that they accept an optional etag and do the conditional update.  In an ideal world, we'd issue a SQL "update" with the etag condition in it (rather than read-check-write).
- We need to plumb that optional etag argument out to all the PUT entrypoints.

That doesn't sound so bad.  We might need better transaction support (see #973) -- I'm not sure.